src/kernel.ts

```diff
@@ -1,1069 +1,803 @@
+  }
+
+  /**
+   * Get info on an object.
+   *
+   * Returns a KernelFuture that will resolve to a `inspect_reply` message documented
+   * [here](http://ipython.org/ipython-doc/dev/development/messaging.html#object-information)
+   */
+  inspect(code: string, cursor_pos: number): IKernelFuture {
+    var content = {
+      code: code,
+      cursor_pos: cursor_pos,
+      detail_level: 0
+    };
+    return this.sendShellMessage("inspect_request", content);
+  }
+
+  /**
+   * Execute given code into kernel, returning a KernelFuture.
+   *
+   * @example
+   *
+   * The options object should contain the options for the execute
+   * call. Its default values are:
+   *
+   *      options = {
+   *        silent : true,
+   *        user_expressions : {},
+   *        allow_stdin : false,
+            store_history: false
+   *      }
+   *
+   */
+  execute(code: string, options?: { silent?: boolean; user_expressions?: any; allow_stdin?: boolean; store_history?: boolean; }): IKernelFuture {
+    var content = {
+      code: code,
+      silent: true,
+      store_history: false,
+      user_expressions: {},
+      allow_stdin: false
+    };
+    utils.extend(content, options);
+    return this.sendShellMessage("execute_request", content);
+  }
+
+  /**
+   * Request a code completion from the kernel.
+   *
+   * Returns a KernelFuture with will resolve to a `complete_reply` documented
+   * [here](http://ipython.org/ipython-doc/dev/development/messaging.html#complete)
+   */
+  complete(code: string, cursor_pos: number): IKernelFuture {
+    var content = {
+      code: code,
+      cursor_pos: cursor_pos
+    };
+    return this.sendShellMessage("complete_request", content);
+  }
+
+  /**
+   * Send an input reply message to the kernel.
+   *
+   * TODO: how to handle this?  Right now called by
+   * ./static/notebook/js/outputarea.js:827:        this.events.trigger('send_input_reply.Kernel', value);
+   *
+   * which has no referense to the session or the kernel
+   */
+  sendInputReply(input: any): string {
+    if (!this.isConnected) {
+      throw new Error("kernel is not connected");
+    }
+    var content = {
+      value: input
+    };
+    var msg = this._getMsg("input_reply", content);
+    msg.channel = 'stdin';
+    this._ws.send(serialize.serialize(msg));
+    return msg.header.msgId;
+  }
+
+  /**
+   * Create a Kernel Message given input attributes.
+   */
+  private _getMsg(msg_type: string, content: any,
+    metadata = {}, buffers: string[] = []): IKernelMsg {
+    var msg: IKernelMsg = {
+      header: {
+        msgId: utils.uuid(),
+        username: this._username,
+        session: this._sessionId,
+        msgType: msg_type,
+        version: "5.0"
+      },
+      metadata: metadata || {},
+      content: content,
+      buffers: buffers || [],
+      parentHeader: {}
+    };
+    return msg;
+  }
+
+  /**
+   * Handle a kernel status change message.
+   */
+  private _handleStatus(status: string) {
```

Comment:
Sorry, that should have been "can be read...". If external code wants to query the current status of the kernel, it is forced to subscribe to the signal and maintain the state itself (but has no way of getting the initial state). My suggestion is for the kernel to store the status internally, and expose it as a read-only property *in addition* to emitting the `statusChanged` signal.